### PR TITLE
Add snowflake-query worker tool example

### PR DESCRIPTION
**Pull request by atvaccaro**
## What

Adds `examples/workers/tools/snowflake-query/` — a Notion worker that lets a custom agent query a Snowflake warehouse read-only. Three tools an agent can chain (find a table → check its columns → query it):

- `listTables` — `SHOW TABLES` (optional database/schema/`LIKE`)
- `describeTable` — `DESCRIBE TABLE`
- `query` — a single read-only `SELECT` / `WITH ... SELECT`

Key-pair (JWT) auth, `node-sql-parser` SELECT-only validation, bounded results + statement timeout. The README walks a customer through a dedicated read-only role + key-pair setup, deploy, secrets, and connecting to a custom agent.

## Verification

- `npm run check` (typecheck), prettier, and markdownlint all clean
- 30/30 credential-free SQL-guard logic tests (accepts SELECT/CTE; rejects DELETE/UPDATE/INSERT/DROP/CREATE/MERGE/multi-statement/SHOW/DESCRIBE/empty/garbage; identifier-injection blocked)
- End-to-end against a live read-only Snowflake connection via `ntn workers exec --local`: `query` (`CURRENT_DATE`), `listTables` (170 tables), `describeTable` (61 columns); a `DELETE` is rejected before reaching Snowflake
- Deployed to the dev workspace; tools register as `listTables` / `describeTable` / `query`

No secrets committed — `.env`, `workers.json`, and key files are gitignored; only `.env.example` (placeholders) ships.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: examples/workers/tools/snowflake-query/src/snowflake.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+          return
+        }
+        const rawRows = (rows ?? []) as Record<string, unknown>[]
+        const columns = (stmt.getColumns?.() ?? []).map((column) => ({
+          name: column.getName?.() ?? "",
+          type: column.getType?.() ?? null,
+          nullable: column.isNullable?.() ?? null,
+        }))
+        resolve({
+          rows: rawRows.map(normalizeRow),
```
Comment (ksinder) [inline examples/workers/tools/snowflake-query/src/snowflake.ts:139]: **[Non-blocking] Duplicate output column names collapse silently in `rows`**

The Snowflake Node SDK returns each row as an object keyed by column name, so when a query yields two columns with the same name (easy to hit with `SELECT a.*, b.*` joins that both have `id` or `created_at`), one value is dropped before `normalizeRow` runs. `columns` on line 130 still lists both, so the agent gets more columns than per-row keys, with no error. Fine to leave for an example; to close it, read rows with `rowMode: "array"` and zip against `columns`. (ref: snowflakedb/snowflake-connector-nodejs#114)

File: examples/workers/tools/snowflake-query/src/sql.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// and string literals can't smuggle in a write. This is a convenience check,
+// not the security boundary -- that's the read-only Snowflake role (see README).
+export function assertSelectOnly(sql: string): void {
+  if (!sql.trim()) {
+    throw new Error("sql must be a non-empty string.")
+  }
+
+  let statements
+  try {
+    statements = parser.astify(sql, { database: "Snowflake" })
```
Comment (ksinder) [inline examples/workers/tools/snowflake-query/src/sql.ts:15]: **[Non-blocking] The parser also rejects some valid read-only queries**

node-sql-parser's Snowflake dialect is alpha. I checked the common reads against 5.3.x and they parse fine (`QUALIFY`, `::` casts, `col:field` access, CTEs, `LATERAL FLATTEN`), but `SAMPLE` (e.g. `SELECT * FROM t SAMPLE (10)`) throws and gets rejected. It fails closed, so it's safe. The README's safety note only mentions writes/DDL/multi-statement, so one line saying a few legitimate reads can bounce too would set expectations.

**Review by ksinder — APPROVED** [review]
Approving. Clean, well-documented example, and the safety story is the right one: the README is explicit that the read-only Snowflake role, not the SQL check, is the real boundary. That's exactly what people copying this should take away.

**What I verified**

- Ran the SQL guard against node-sql-parser's Snowflake dialect directly. Writes are rejected: `DELETE`/`UPDATE`/`INSERT`/`TRUNCATE`/`CALL` parse to non-select types, `MERGE`/`COPY INTO` throw, multi-statement is caught by the length check, and `SELECT ... INTO` (which parses *as* a select) is caught by your `into.expr` guard. Nothing wrote through.
- Secrets hygiene is right: `.env`/`workers.json`/`*.p8`/`rsa_key*` gitignored, only placeholder `.env.example` ships, secret scanner green.
- CI green (Pi scan still pending, not a blocker).

Two small non-blocking notes inline, both fail-safe edge cases. One process note: the PR body cites 30/30 SQL-guard tests, but no test file is in the PR. Since `sql.ts` is the security-relevant bit, committing those tests would keep the guard honest across future edits. Not a blocker, and examples CI only runs typecheck + markdownlint today anyway.

[review written by Claude Code on behalf of @ksinder]

**Review by ksinder — APPROVED** [review]

**Comment by atvaccaro** [comment]
### Merge activity

* **Jun 25, 11:28 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/makenotion/notion-cookbook/22).
* **Jun 25, 11:28 PM UTC**: [Graphite](https://app.graphite.com/github/pr/makenotion/notion-cookbook/22) couldn't merge this PR because it failed for an unknown reason (You're not authorized to push to this branch).
